src/coreclr/jit/codegenloongarch64.cpp

```diff
@@ -0,0 +1,9260 @@
+
+        // first we load the immediate into tmpReg
+        assert(!EA_IS_RELOC(size));
+        GetEmitter()->emitIns_I_la(size, tmpReg, imm);
+        regSet.verifyRegUsed(tmpReg);
+
+        // when we are in an unwind code region
+        // we record the extra instructions using unwindPadding()
+        if (inUnwindRegion)
+        {
+            compiler->unwindPadding();
+        }
+
+        if (ins == INS_addi_d)
+        {
+            GetEmitter()->emitIns_R_R_R(INS_add_d, attr, reg1, reg2, tmpReg);
+        }
+        else
+        {
+            GetEmitter()->emitIns_R_R_R(INS_add_d, attr, tmpReg, reg2, tmpReg);
+            GetEmitter()->emitIns_R_R_I(ins, attr, reg1, tmpReg, 0);
+        }
+    }
+    return immFitsInIns;
+}
+
+//------------------------------------------------------------------------
+// genStackPointerAdjustment: add a specified constant value to the stack pointer in either the prolog
+// or the epilog. The unwind codes for the generated instructions are produced. An available temporary
+// register is required to be specified, in case the constant is too large to encode in an "daddu"
+// instruction (or "dsubu" instruction if we choose to use one), such that we need to load the constant
+// into a register first, before using it.
+//
+// Arguments:
+//    spDelta                 - the value to add to SP (can be negative)
+//    tmpReg                  - an available temporary register
+//    pTmpRegIsZero           - If we use tmpReg, and pTmpRegIsZero is non-null, we set *pTmpRegIsZero to 'false'.
+//                              Otherwise, we don't touch it.
+//    reportUnwindData        - If true, report the change in unwind data. Otherwise, do not report it.
+//
+// Return Value:
+//    None.
+
+void CodeGen::genStackPointerAdjustment(ssize_t spDelta, regNumber tmpReg, bool* pTmpRegIsZero, bool reportUnwindData)
+{
+    // Even though INS_addi_d is specified here, the encoder will choose either
+    // an INS_add_d or an INS_addi_d and encode the immediate as a positive value
+    //
+    bool wasTempRegisterUsedForImm =
+        !genInstrWithConstant(INS_addi_d, EA_PTRSIZE, REG_SPBASE, REG_SPBASE, spDelta, tmpReg, true);
+    if (wasTempRegisterUsedForImm)
+    {
+        if (pTmpRegIsZero != nullptr)
+        {
+            *pTmpRegIsZero = false;
+        }
+    }
+
+    if (reportUnwindData)
+    {
+        // spDelta is negative in the prolog, positive in the epilog,
+        // but we always tell the unwind codes the positive value.
+        ssize_t  spDeltaAbs    = abs(spDelta);
+        unsigned unwindSpDelta = (unsigned)spDeltaAbs;
+        assert((ssize_t)unwindSpDelta == spDeltaAbs); // make sure that it fits in a unsigned
+
+        compiler->unwindAllocStack(unwindSpDelta);
+    }
+}
+
+//------------------------------------------------------------------------
+// genPrologSaveRegPair: Save a pair of general-purpose or floating-point/SIMD registers in a function or funclet
```

Comment:
> But, if I understand correctly, LA does not have register pair save/restore. So this is a lot of unnecessary complexity. Can you just do something simpler?

You are right for current LoongArch64's ISA.
We just keep these code for two reasons:
(1) the previous mips64 which we had extend ISA supports paired load/store;
(2) LA64 is considing add is in future, left these is just for future extention.


> Also, it appears LA does not have pre-index or post-index addressing modes like arm64, and that is another big contributor to prolog/epilog complexity. For example, we try to use pre-indexed addressing to avoid needing a separate "sub sp" instruction. But LA always will require the "sub sp", so there's no need to optimize. This is one of the main reasons there are so many arm64 "frame types" -- presumably LA doesn't need all that complexity and can use a much simpler set of "frame types", preferably just one.

Agree with you.
But I think there is a necessary to keep different "frame types" for the large frame size which offset is out range one instruction's offset.

> For instance, maybe you can always just store FP/RA at the top of the frame, and don't need `IsSaveFpRaWithAllCalleeSavedRegisters()` at all -- it will always be true.

Yes, `IsSaveFpRaWithAllCalleeSavedRegisters()` is always true.
The position of FP/RA can be fixed.

I will amend it later and will discuss with you.

Reply:
@BruceForstall 
Can I keep this implementation for LA  which can be refactored in future?